src/GridConnect.cpp

```diff
@@ -0,0 +1,244 @@
+//     Identifier and data fields are treated as base-16 digits (hexadecimal).
+//     'S' & 'X' characters indicates standard or extended CAN frames
+//     'N' character indicates a normal (not RTR) message
+
+// the GridConnect message syntax for a 'Request to Transmit' (RTR) message (there is no data field)
+// : <S | X> <IDENTIFIER> <R> <LENGTH>;
+// where:
+//     ':' and ';' are message start and end characters respectively
+//     Identifier field is treated as base-16 digits (hexadecimal).
+//     'S' & 'X' characters indicates standard or extended CAN frames
+//     'R' character indicates an RTR message
+//     Length field is a single ASCII decimal character from ‘0’ to ‘8’ that specifies the length
+//     of the message
+
+// the difference in the GridConnect version used here is in the format of the Identifier field:
+//
+// For a standard can message with an 11 bit identifier, the GridConnect identifier field is 4 hex characters,
+// i.e. 16 bits
+// The CBUS implementation shifts the 11 bit ID by 5, i.e. fills the top bit, with the lower 5 bits being 0
+// (This was done to map directly onto SIDH and SIDL registers in the PIC processor family)
+// And the GridConnect Identifier field is also leading zero padded, so always 4 characters for a standard message
+// 
+// For an extended can message with an 29 bit identifier, the GridConnect identifier field is 8 hex characters,
+// i.e. 32 bits
+// In this case the mapping is a bit more complex
+// IDENTIFIER[0,1] - bits [28:21] of the can identifier
+// IDENTIFIER[2] - bits [20:18] of the can identifier, left shifted by 1, bottom bit set to zero
+// IDENTIFIER[3] - bits [17:16] of the can identifier, bottom two bits, top two bits set to 0
+// IDENTIFIER[4,5,6,7] - bits [15:0] of the can identifier
+// (This maps directly onto SIDH, SIDL, EIDH and EIDL registers the PIC processor family)
+// And the GridConnect Identifier field is also leading zero padded, so always 8 characters for an extended message
+// 
+
+#include <ctype.h>
+#include "GridConnect.h"
+
+namespace VLCB
+{
+
+  bool encodeGridConnect(char * gcBuffer, CANMessage *msg){
+      byte offset = 0;
+      gcBuffer[0] = 0;  // null terminate buffer to start with
+      // set starting character & standard or extended CAN identifier
+      if (msg->ext) {
+        if (msg->id > 0x1FFFFFFF)
+        {
+          // id is greater than 29 bits, so fail the encoding
+          return false;
+        }
+        // mark as extended message
+        strcpy (gcBuffer,":X");
+        // extended 29 bit CAN idenfier in bytes 2 to 9
+        // chars 2 & 3 are ID bits 21 to 28
+        sprintf(gcBuffer + 2, "%02X", (msg->id) >> 21);
+        // char 4 -  bits 1 to 3 are ID bits 18 to 20
+        sprintf(gcBuffer + 4, "%01X", ((msg->id) >> 17) & 0xE);
+        // char 5 -  bits 0 to 1 are ID bits 16 & 17
+        sprintf(gcBuffer + 5, "%01X", ((msg->id) >> 16) & 0x3);
```

Comment:
Are these bits correct? Looks like the 29 bits are not contiguous in the hex string.

Reply:
Thats correct - thats the difference between the 'official' GridConnect and the 'CBUS' variant of Grid connect
The hex values are mangled to fit the can controller used in PIC's, so for 11 bit ID's the lower 5 bits aren't used, and for 29 bit ID's there are three unused bits in the 3rd byte
I've tried to show this in the notes at the start of the file - but I've found it hard to describe how it's split
That's why the unit tests where so important, to be able to test the boundaries of this split - I think the unit tests will show this split better than words

Reply:
Yes, the unit test explains it all.